src/symbolize/gimli/elf.rs

```diff
@@ -10,11 +15,150 @@ type Elf = object::elf::FileHeader32<NativeEndian>;
 #[cfg(target_pointer_width = "64")]
 type Elf = object::elf::FileHeader64<NativeEndian>;
 
+const BUILD_ID_PATH: &[u8] = b"/usr/lib/debug/.build-id/";
+const BUILD_ID_SUFFIX: &[u8] = b".debug";
+
 impl Mapping {
     pub fn new(path: &Path) -> Option<Mapping> {
         let map = super::mmap(path)?;
-        Mapping::mk(map, |data, stash| Context::new(stash, Object::parse(data)?))
+        let object = Object::parse(&map)?;
+
+        // Try to locate an external debug file using the build ID.
+        let build_id = object.build_id().unwrap_or(&[]);
+        if !build_id.is_empty() {
+            fn hex(byte: u8) -> u8 {
+                if byte < 10 {
+                    b'0' + byte
+                } else {
+                    b'a' + byte - 10
+                }
+            }
+            let mut path = Vec::with_capacity(
+                BUILD_ID_PATH.len() + BUILD_ID_SUFFIX.len() + build_id.len() * 2 + 1,
+            );
+            path.extend(BUILD_ID_PATH);
+            path.push(hex(build_id[0] >> 4));
+            path.push(hex(build_id[0] & 0xf));
+            path.push(b'/');
+            for byte in &build_id[1..] {
+                path.push(hex(byte >> 4));
+                path.push(hex(byte & 0xf));
+            }
+            path.extend(BUILD_ID_SUFFIX);
+            if let Some(mapping) = Mapping::new_debug(Path::new(OsStr::from_bytes(&path)), None) {
+                return Some(mapping);
+            }
+        }
+
+        // Try to locate an external debug file using the GNU debug link section.
+        let tmp_stash = Stash::new();
+        if let Some(section) = object.section(&tmp_stash, ".gnu_debuglink") {
+            if let Some(len) = section.iter().position(|x| *x == 0) {
+                let filename = &section[..len];
+                let offset = (len + 1 + 3) & !3;
+                if let Some(crc_bytes) = section
+                    .get(offset..offset + 4)
+                    .and_then(|bytes| bytes.try_into().ok())
+                {
+                    let crc = u32::from_ne_bytes(crc_bytes);
+                    if let Some(path_debug) = locate_debuglink(path, filename) {
+                        if let Some(mapping) = Mapping::new_debug(&path_debug, Some(crc)) {
+                            return Some(mapping);
+                        }
+                    }
+                }
+            }
+        }
+
+        let stash = Stash::new();
+        let cx = Context::new(&stash, object)?;
+        Some(Mapping {
```

Comment:
This no longer uses `Mapping::mk` because it needs to parse the `Object` beforehand. Not sure if there is any way to restructure this to allow that.